Frameworks/Foundation/NSProxy.mm

```diff
@@ -14,220 +14,317 @@
+    return aClass == forwardClass;
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (BOOL)respondsToSelector:(SEL)aSelector {
-    UNIMPLEMENTED();
-    return StubReturn();
+    return [self methodSignatureForSelector:aSelector] != nil;
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (BOOL)conformsToProtocol:(Protocol*)aProtocol {
-    UNIMPLEMENTED();
-    return StubReturn();
+    Class forwardClass = static_cast<Class>([self performSelector:@selector(class)]);
+    return class_conformsToProtocol(forwardClass, aProtocol);
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (id)performSelector:(SEL)aSelector {
-    UNIMPLEMENTED();
-    return StubReturn();
+    NSInvocation* invocation = [NSInvocation invocationWithMethodSignature:[self methodSignatureForSelector:aSelector]];
+    invocation.selector = aSelector;
+    [self forwardInvocation:invocation];
+
+    id ret;
+    [invocation getReturnValue:&ret];
+    return ret;
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (id)performSelector:(SEL)aSelector withObject:(id)anObject {
-    UNIMPLEMENTED();
-    return StubReturn();
+    NSInvocation* invocation = [NSInvocation invocationWithMethodSignature:[self methodSignatureForSelector:aSelector]];
+    invocation.selector = aSelector;
+    [invocation setArgument:anObject atIndex:2];
+    [self forwardInvocation:invocation];
+
+    id ret;
+    [invocation getReturnValue:&ret];
+    return ret;
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (id)performSelector:(SEL)aSelector withObject:(id)anObject withObject:(id)anotherObject {
-    UNIMPLEMENTED();
-    return StubReturn();
+    NSInvocation* invocation = [NSInvocation invocationWithMethodSignature:[self methodSignatureForSelector:aSelector]];
+    invocation.selector = aSelector;
+    [invocation setArgument:anObject atIndex:2];
+    [invocation setArgument:anotherObject atIndex:3];
+    [self forwardInvocation:invocation];
+
+    id ret;
+    [invocation getReturnValue:&ret];
+    return ret;
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (BOOL)isProxy {
-    UNIMPLEMENTED();
-    return StubReturn();
+    return YES;
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (instancetype)retain {
-    UNIMPLEMENTED();
-    return StubReturn();
+    intptr_t* refCount = ((intptr_t*)self) - 1;
+    // Note: this should be an atomic read, so that a sufficiently clever
+    // compiler doesn't notice that there's no happens-before relationship
+    // here.
+    if (*refCount >= 0) {
+        __sync_add_and_fetch(refCount, 1);
+    }
+    return self;
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (oneway void)release {
-    UNIMPLEMENTED();
+    intptr_t* refCount = ((intptr_t*)self) - 1;
+    // We allow refcounts to run into the negative, but should only
+    // deallocate once.
+    if (__sync_sub_and_fetch(refCount, 1) == -1) {
```

Comment:
- this is just lifted from NSObject.mm
- async nonsense wouldn't be able to cause that since __sync_sub_and_fetch is atomic